### PR TITLE
[#11641] Apply lazy-loading of question responses in session results page

### LINKED ISSUE #11641: Apply lazy-loading of question responses in session results page
Current: Upon loading the session results page, the responses of all questions are loaded.

Problems:
- Not performant at large number of responses.
  - While it does not introduce burst of parallel API calls, as the burden of fetching all responses falls on one single API call, it is not difficult to see the potential performance issue at large enough scale.
- There is a not-small chance that a user only wants to see the responses of a question somewhere before or until the middle of the page. That means the bottom half of the page is wasted.

Note: While this is in the same spirit as #11640, the root cause is different and more fundamental here.

Solution:
- Upon loading the page, only fetch the list of questions, just like how submission page and feedback report page for instructors work.
- (similar to #11640) Apply lazy loading; only load the responses of the question when the question is in viewport.

### PR DESCRIPTION
**Pull request by tsy24**
Fixes #11641 

Ready for review

**Outline of Solution**
Solution of loading the responses of the question when the question is in viewport is chosen over question tabs similar to the feedback report page for instructors:
- Standardize interface for students: loading when question in viewport is similar to the student submission page
- Stick to the current behaviour of not showing questions with no response

For session result, load responses of each question individually:
- Load list of questions first with `getFeedbackQuestions`
- For each question, load response when question is in viewport with `getFeedbackSessionResults`
- `FeedbackQuestionModel` to model a question on the session result page
    - `isLoaded` for whether the responses of a question has been loaded
    - `isLoading` for whether the responses of a question is currently loading
    - `hasResponse` for whether there are responses to be displayed for that question

After applying lazy loading:
image

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/GetFeedbackQuestionsAction.java [hunk 1]
```diff
@@ -36,7 +36,11 @@ void checkSpecificAccessControl() throws UnauthorizedAccessException {
             break;
         case FULL_DETAIL:
             gateKeeper.verifyLoggedInUserPrivileges(userInfo);
-            gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(courseId, userInfo.getId()), feedbackSession);
+            if (userInfo.isStudent) {
+                gateKeeper.verifyAccessible(getStudentOfCourseFromRequest(courseId), feedbackSession);
+            } else {
+                gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(courseId, userInfo.getId()), feedbackSession);
+            }
             break;
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/GetFeedbackQuestionsAction.java:44]: Could you elaborate on why is this change required? I don't think students need to see full detail.

Comment (tsy24) [reply src/main/java/teammates/ui/webapi/GetFeedbackQuestionsAction.java]: Have reverted the change after looking at it again. Instead of using `FULL_DETAIL`, I have made changes so that `STUDENT_RESULT` intent can be used.

File: src/web/app/components/question-response-panel/question-response-panel.component.ts [hunk 2]
```diff
@@ -64,4 +75,39 @@ export class QuestionResponsePanelComponent {
...
+          question.feedbackQuestion = responses.feedbackQuestion;
+          question.allResponses = responses.allResponses;
+          question.otherResponses = responses.otherResponses;
+          question.questionStatistics = responses.questionStatistics;
+          question.responsesFromSelf = responses.responsesFromSelf;
+          question.responsesToSelf = responses.responsesToSelf;
+        } else {
+          question.hasResponse = false;
+        }
+      }, (resp: ErrorMessageOutput) => this.statusMessageService.showErrorToast(resp.error.message));
```
Comment (samuelfangjw) [inline src/web/app/components/question-response-panel/question-response-panel.component.ts:104]: Current behavior when error occurs, error toast is shown, question is marked as loaded and "No responses received yet" is shown" or similar message is shown. 

Screen Shot 2022-07-31 at 11 28 29 AM
 
Need to make sure the error message persists similar to https://github.com/TEAMMATES/teammates/pull/11863#issuecomment-1167405928.

File: src/web/app/components/question-response-panel/question-response-panel.component.ts [hunk 3]
```diff
@@ -64,4 +75,39 @@ export class QuestionResponsePanelComponent {
     return false;
   }
 
+  /**
+   * Loads responses for feedback question.
+   */
+  loadQuestionResults(question: FeedbackQuestionModel): void {
```
Comment (samuelfangjw) [inline src/web/app/components/question-response-panel/question-response-panel.component.ts:81]: Can this method be private?

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/e2e/java/teammates/e2e/pageobjects/FeedbackResultsPage.java [hunk 4]
```diff
@@ -399,6 +399,9 @@ private StringBuilder appendOptions(StringBuilder info, List<String> options) {
     }
 
     private WebElement getQuestionResponsesSection(int questionNum) {
+        WebElement question = browser.driver.findElement(By.id("question-" + questionNum + "-responses"));
+        scrollElementToCenter(question);
+        waitUntilAnimationFinish();
         return browser.driver.findElement(By.id("question-" + questionNum + "-responses"));
```
[the remark this answers is not in the document]
Comment (tsy24) [reply src/e2e/java/teammates/e2e/pageobjects/FeedbackResultsPage.java:405]: In my opinion, it is necessary to find the element again. This is due to the behaviour that questions with no responses will not be shown. 
As the questions are now loaded individually, the question will be present initially before the responses are loaded. Thus `WebElement question = browser.driver.findElement(By.id("question-" + questionNum + "-responses"));` will not throw any exception. If the method simply returns `question`, no exception is caught in `verifyQuestionNotPresent()` and the test case will fail. 
By using `return browser.driver.findElement(By.id("question-" + questionNum + "-responses"));`, the question's responses would have been loaded by then, and questions with no responses will not be present, causing an exception to be thrown and the test case will then pass.

**Comment by hhdqirui** [comment]
Hi @tsy24 reminder on this PR.

**Comment by tsy24** [comment]
Changes have been made based on changes requested. PR is ready for review.

**Comment by fsgmhoward** [comment]
@samuelfangjw @jayasting98 Reminder for this PR review.

**Review by samuelfangjw — APPROVED** [review]

**Comment by zhaojj2209** [comment]
@NicolasCwy Do help to give this PR a final review, thanks!
@tsy24 Do remember to remove the extra comment as well.

**Review by NicolasCwy — APPROVED** [review]
LGTM !
